Add new Tutorial - Using Ludwig Experiment to build Image Classifier from MNIST dataset (#5333)
galaxyproject/training-material: topics/statistics/tutorials/galaxy-ludwig/tutorial.md

![schema of the whole process of training model and test.](../../images/galaxy-ludwig/explain_model_schema.png "Overview of the steps process to obtain the handwritten classification model and testing it.")

> <agenda-title></agenda-title>
>
> In this tutorial, we will cover:
>
> 1. TOC
> {:toc}
>
{: .agenda}

> <comment-title>Background</comment-title>
>
> [MNIST dataset](https://yann.lecun.com/exdb/mnist/) provides the database of handwritten digits
> the database has a training set of 60,000 examples, and a test set of 10,000 examples.
> at our planet and its environment for the ultimate benefit of all European citizens.
> It is a subset of a larger set available from NIST.
> The digits have been size normalized and centered in a fixed-size image.
{:  .comment}

# Files Format
Before starting our hands-on, here is a brief explanation of the three files generated for the Ludwig Experiment tool.

> ## Image_Files.zip 

Reviewer: Do we need to have `.zip` in the header of this section? Can we write `Images` or `Training images` or something similar? Does `.zip` has some significance? Similar comment for `MNIST_dataset.csv`

Author: Your point is valid. 
I removed the files extension  from the header and changed the titles.

The MNIST dataset consists of images and their corresponding labels. For the purpose of this tutorial, mnist_dataset.csv file is created and contains three columns: image_path, label and, split.

Briefly, the image_path column provides the file paths to the images that will be fed into the deep learning algorithm. The label column contains the correct classifications, ranging from 0 to 9, for the handwritten digits in the images. The split column indicates whether the data should be used for training (0) or testing (2) the model.

![Dataset.csv file format snapshot](../../images/galaxy-ludwig/explain_dataset_format.png "Dataset.csv file format snapshot. features in order: file_path, label, split."){: width="50%" height="10%"}
 
## Config.yaml

The config.yaml file is crucial as it defines the entire structure of your machine learning experiment. This configuration file tells Ludwig how to process your data, what model to use, how to train it, and what outputs to generate.

The rationale on how this file was constructed for this dataset is the following:
i) The model takes images as input and uses a stacked convolutional neural network (CNN) to extract features.

> ii) It consists of two convolutional layers followed by a fully connected layer, with dropout applied to both the second convolutional layer and the fully connected layer to reduce overfitting. 

Reviewer: Do we have other layers such as `maxpooling` and `normalisation` used in the architecture of stacked CNN? If yes, please mention them.

iii) The split configuration is used to divide the dataset based on the 'split' column rather than using a random split (which is the default). In this configuration, rows labeled with '0' are included in the training set, while rows labeled with '2' are included in the testing set. If a validation category is needed, rows are labeled with '1'. 
iv) The model is trained to classify images into categories (e.g., recognizing digits), and it will train for 5 epochs.

![Config.yaml file snapshot](../../images/galaxy-ludwig/explain_config.png "Config.yaml file snapshot."){: width="50%" height="5%"}

# Galaxy-Ludwig Tool

> Ludwig simplifies the complexities of machine learning by automating essential steps such as data preprocessing, model architecture selection, hyperparameter tuning, and device management. This streamlined approach is particularly beneficial for Galaxy users who are more interested in addressing their scientific questions than in navigating the intricacies of machine learning workflows.

Reviewer: To automatically do `data preprocessing, model architecture selection, hyperparameter tuning, and device management`, what parameters should a user consider changing/updating based on a different dataset?

Author: I added some lines explaining and providing examples of modifications users can make to the configuration to better fit their dataset and produce a specific model.

> <tip-title>Overfitting</tip-title>
>Overfitting occurs when the network fails to generalize well to unseen data because it begins to model the noise and randomness in the training dataset.
>![alt](../../images/galaxy-ludwig/explain_overfitting.png "Overfitting curve")
>
{: .tip}

> <tip-title>Underfitting</tip-title>
>Underfitting happens when the patterns in the training dataset are too complex for our model. It can’t capture them, so it has a poor performance on training and validation sets. 
>![alt](../../images/galaxy-ludwig/explain_underfitting.png "Underfitting curve")
>
{: .tip}

![loss vs epochs](../../images/galaxy-ludwig/test1_combined_loss.png "Loss versus Epochs during training. The graph illustrates the relationship between the training loss and the number of epochs during the training of the model.")

Based on the training curve (blue line), we can draw the following conclusions:

- CONSISTENT IMPROVEMENT:

>     Training loss consistently decrease over the five epochs. This indicates that the model is effectively learning and improving its performance on the dataset.

Reviewer: decreases

Author: Thank you. Fixed.

- APPROACHING CONVERGENCE:

>     By Epoch 5, the training loss has reduced to less than 0.5. The gradual reduction in losses suggests that the model is approaching convergence, where further training may yield diminishing returns in terms of performance improvement.

Reviewer: Do we know something about the `test loss`? Observing test/validation loss is also important to decide whether a model is generalising well or not

Author: Yes, we have this information. 
It is present in the report file under predictions.shapes. 
It is not included in the graph, which is why it isn’t mentioned directly; the focus was to provide a brief explanation of the graphs.

Author: I added an additional information point to highlight this result and its conclusion.

As training progresses, the accuracy on the training data usually increases, indicating that the model is learning to correctly classify more examples. 

> <tip-title>Plotting accuracy and/or Loss</tip-title>
> Plotting accuracy and plotting loss are both essential for monitoring the performance of a deep learning model, but they serve different purposes and provide different insights. While accuracy provides a straightforward measure of model performance, loss offers deeper insights into the training process, making both plots valuable for different aspects of model evaluation.
{: .tip}

![alt](../../images/galaxy-ludwig/test1_label_accuracy.png "Accuracy versus Epochs. This graph depicts the accuracy of the model over the course of training. The training accuracy curve indicates how well the model is performing on the training data.")

Based on the training curve (blue line), we can draw the following conclusions:

- THE DATASET IS RELATIVELY EASY FOR THE MODEL TO LEARN:

>     The model starts with a relatively high training accuracy of >0.8.

Reviewer: Accuracy on test data should also be mentioned or cross-validation accuracy to decide a model really learns well.

Author: I added an additional information point to highlight this result and its conclusion.

---
layout: tutorial_hands_on
level: Intermediate
title: Train and Test a Deep learning image classifier with Galaxy-Ludwig
questions:
- How do I use the Ludwig tool to create and run a deep learning image classifier? 
- What files do I need to run Ludwig? 
- How do I build and evaluate the image classification model? 
objectives:
- Execute a Ludwig experiment. 
- Evaluate the image classfier model. 
- Compare different configurations to improve model performance. 
time_estimation: 40M
key_points:
- Use Galaxy-Ludwig Experiment to solve an image classification problem
- Using CNN on MNIST dataset. 
contributors:
- paulocilasjr 
- qchiujunhao 
- jgoecks 
tags:
- MNIST
- Deep learning
- Ludwig
---

> <comment-title>Galaxy-Ludwig Tool</comment-title>
>
> The Ludwig tool described in this tutorial is only available at: 

> > ```https://cancer.usegalaxy.org/```

Reviewer 2: Any reason for this? I don't see the tool in the tool shed either, are there plans to add this? Adding the tool to IUC would be ideal, so that other Galaxies can also support this tutorial.

Reviewer 2: secondly, you probably want that url to be a proper hyperlink, not a comment

Author: Tomorrow, I'm going to have a more precise answer regarding the point you raised.

Author: I added the toll in the tools shed (https://toolshed.g2.bx.psu.edu/view/paulo_lyra_jr/ludwig_applications/3e565bbe8b71)

Not sure we can get this tool (+ original Galaxy-ML tools) into the IUC yet.

{:  .comment}

Using the MNIST image dataset of handwritten digits as input, we will build an image recognition model with the Galaxy-Ludwig tool. The goal is to classify the handwritten digit in each image.

To accomplish this, three steps are needed: (i) upload Ludwig files and image files to Galaxy (ii) Set up and running the Ludwig experiment function on Galaxy, and (iii) Evaluate the image classification model. As a bonus step, we'll also explore (iv) improving the model's classification performance (Figure 1).

![schema of the whole process of training model and test.](../../images/galaxy-ludwig/explain_model_schema.png "Overview of the steps process to obtain the handwritten classification model and testing it.")

> <!-- You may want to cite some publications; this can be done by adding citations to the

Reviewer 2: please feel free to remove these autogenerated hints

Author: Thank you. Just removed them.

bibliography file (`tutorial.bib` file next to your `tutorial.md` file). These citations
must be in bibtex format. If you have the DOI for the paper you wish to cite, you can
get the corresponding bibtex entry using [doi2bib.org](https://doi2bib.org).

With the example you will find in the `tutorial.bib` file, you can add a citation to
this article here in your tutorial like this:
{% raw %} `{% cite Batut2018 %}`{% endraw %}.
This will be rendered like this: {% cite Batut2018 %}, and links to a
[bibliography section](#bibliography) which will automatically be created at the end of the
tutorial.-->

> <agenda-title></agenda-title>
>
> In this tutorial, we will cover:
>
> 1. TOC
> {:toc}
>
{: .agenda}

> <comment-title>Background</comment-title>
>
> [MNIST dataset](https://yann.lecun.com/exdb/mnist/) provides the database of handwritten digits
> the database has a training set of 60,000 examples, and a test set of 10,000 examples.
> at our planet and its environment for the ultimate benefit of all European citizens.
> It is a subset of a larger set available from NIST.
> The digits have been size normalized and centered in a fixed-size image.
{:  .comment}

> # FILES FORMAT

Reviewer 2: please don't use all caps for the section headings

Author: No problem! I will change them all.

Since our model will learn directly from the PNG files containing the handwritten numbers, we need to upload these files as well. The images are stored in two main folders: one for training and one for testing. Another folder, named after the figure label, is used to store the figures (Figure 2).

Notice that we are going to work with compressed file (.zip), Galaxy-Ludwig knows how to decompress the files and have them ready.

![folder tree related to the hand-written digits picture](../../images/galaxy-ludwig/explain_images_file.png "File tree containing the images files used to training and testing the model")

## MNIST_dataset.csv

The MNIST dataset consists of images and their corresponding labels. For the purpose of this tutorial, mnist_dataset.csv file is created and contains three columns: image_path, label and, split.

Briefly, the image_path column provides the file paths to the images that will be fed into the deep learning algorithm. The label column contains the correct classifications, ranging from 0 to 9, for the handwritten digits in the images. The split column indicates whether the data should be used for training (0) or testing (2) the model.

> ![Dataset.csv file format snapshot](../../images/galaxy-ludwig/explain_dataset_format.png "Dataset.csv file format snapshot. features in order: file_path, label, split.")

Reviewer 2: Just as a tip, since this image is rendering perhaps larger than you want, you can add some inline styling to the end like so: (have not tested if 50% is a good value, but just as an example)

```suggestion
![Dataset.csv file format snapshot](../../images/galaxy-ludwig/explain_dataset_format.png "Dataset.csv file format snapshot. features in order: file_path, label, split."){: width="50%"}
```

Author: Thank you! This is great to know. I will test it and submit it with the best configuration.

>
>    ```
>    https://zenodo.org/records/13292895/files/mnist_images.zip
>    https://zenodo.org/records/13357260/files/mnist_dataset.csv 
>    https://zenodo.org/records/13713095/files/config.yaml
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
> 3. Check that the data format assigned for each file is **zip**, **csv**, and **yaml**, respectively.
>    If it is not, follow the Changing the datatype tip.
>
>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="datatypes" %}
>
> 5. Add a tag (`Stacked CNN`) to the dataset corresponding to `config.yaml`
>    This is important to trace back on what configurations the model was built on.
>
>    {% snippet faqs/galaxy/datasets_add_tag.md %}
>
{: .hands_on}

## Ludwig Experiment

> <hands-on-title> Task description </hands-on-title>
>
> 1. {% tool [Ludwig Experiment](ludwig_experiment) %} with the following parameters:
>    - {% icon param-file %} *"Select the dataset containing model configuration"*: `config.yaml`
>    - {% icon param-file %} *"Input dataset"*: `mnist_dataset.csv`
>    - {% icon param-file %} *"Raw data"*: `mnist_images.zip`
>

> >    > <comment-title> short description </comment-title>

Reviewer 2: you probably want to update or remove this comment title

Reviewer 2: furthermore, I think we can assume people have enough familiarity with Galaxy to know to push the "RUN TOOL" button, so the box can also just be removed completely

Author: I agree. I removed it all.

>    >
>    > Once all the inputs are in place, press the 'RUN TOOL' button to start your job.
>    {: .comment}
>
{: .hands_on}

## Ludwig Experiment Outputs files

After your model is trained and tested, you should see three new files in your history list:

>     > Ludwig Experiment Report: An HTML file containing the evaluation report of the trained model.

Reviewer 2: I think this could look better as a markdown list rather than comment block

Author: Changed it to list.

# CHANGING MODEL'S CONFIGURATION (Epoch = 15)

When designing a neural network, it is unfortunately not possible to determine the optimal number of neurons, layers, or other hyperparameters solely from a problem description. There is limited guidance available for selecting appropriate initial values, making it challenging to identify a starting point.

A widely adopted approach involves making an initial estimate based on prior experience with networks designed for similar problems. This experience could be derived from your own work or gathered from secondary sources, such as training courses, blogs, or research papers. From there, it is advisable to experiment with different configurations and carefully evaluate the performance of each before selecting the most effective architecture.

Since we have gathered insights from our recent model, we can now begin testing new configurations and comparing them to the previous results.

## Changing Epochs

For this new configuration, the file has been modified to adjust the Epoch value, increasing it from 5 to 15. Upload the file using the following link:

```

>     https://zenodo.org/records/13716147/files/config_test_2.yaml

Reviewer 2: ```suggestion
https://zenodo.org/records/13716147/files/config_test_2.yaml
```

iii) The fully connected layer helps to reduce overfitting.
iv) The split configuration is used to divide the dataset based on the 'split' column rather than using a random split (which is the default). In this configuration, rows labeled with '0' are included in the training set, while rows labeled with '2' are included in the testing set. If a validation category is needed, rows are labeled with '1'.
v) The model is trained to classify images into categories (e.g., recognizing digits) and will train for 5 epochs.

![Config.yaml file snapshot](../../images/galaxy-ludwig/explain_config.png "Config.yaml file snapshot."){: width="50%" height="5%"}

# Galaxy-Ludwig Tool

Ludwig simplifies the complexities of machine learning by automating essential steps such as data preprocessing, model architecture selection, hyperparameter tuning, and device management. To properly adapt the configuration for different datasets, users should consider the following parameters:

- Input Features: Ensure the type reflects the nature of the data.

> - Encoder: Depending on the input type, this may need to be changed. For example, for image data, consider varying num_filters and filter_size. For text data, you might switch to type: rnn or type: transformer.

Reviewer: ```suggestion
- Encoder: Depending on the input type, this may need to be changed. For example, for image data, consider varying `num_filters` and `filter_size` hyperparameters. For text data, you might switch to type: rnn or type: transformer.
```

## Prepare environment and get the data 

> <hands-on-title> Environment and Data Upload </hands-on-title>
>
> 1. Create a new history for this tutorial. If you are not inspired, you can name it *image classifier*.
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Import the files from Zenodo or from the shared data library
>
>    ```
>    https://zenodo.org/records/13292895/files/mnist_images.zip
>    https://zenodo.org/records/13357260/files/mnist_dataset.csv 
>    https://zenodo.org/records/13713095/files/config.yaml
>    ```
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
> 3. Check that the data format assigned for each file is **zip**, **csv**, and **yaml**, respectively.
>    If it is not, follow the Changing the datatype tip.
>
>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="datatypes" %}
>
> 5. Add a tag (`Stacked CNN`) to the dataset corresponding to `config.yaml`
>    This is important to trace back on what configurations the model was built on.
>
>    {% snippet faqs/galaxy/datasets_add_tag.md %}
>
{: .hands_on}

## Ludwig Experiment

> <hands-on-title> Task description </hands-on-title>
>

> > 1. {% tool [Ludwig Experiment](ludwig_experiment) %} with the following parameters:

Reviewer: Adding tool's version would be nice! You can look at other ML tutorials such as deep learning or regression ones for example.

Author: Thank you. I just added.

>>    If it is not, follow the Changing the datatype tip.
>>
>>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="datatypes" %}
>>
>> 5. Add a tag (`Stacked CNN`) to the dataset corresponding to `config.yaml`
>>    This is important to trace back on what configurations the model was built on.
>>
>>    {% snippet faqs/galaxy/datasets_add_tag.md %}
>>
>{: .hands_on}
>
{: .tip}

Run the experiment once again. If you need a refresher on how to set up the Ludwig experiment to run the tool, follow the instructions below.

><tip-title>Ludwig Experiment</tip-title>
>
>><hands-on-title> Task description </hands-on-title>
>>
>> 1. {% tool [Ludwig Experiment](ludwig_experiment) %} with the following parameters:
>> - {% icon param-file %} *"Select the dataset containing model configuration"*: `config.yaml`
>> - {% icon param-file %} *"Input dataset"*: `mnist_dataset.csv`
>> - {% icon param-file %} *"Raw data"*: `mnist_images.zip`
>>
>>> <comment-title> short description </comment-title>
>>>
>>> Once all the inputs are in place, press the 'RUN TOOL' button to start your job.
>>{: .comment}
>>
>{: .hands_on}
>
{: .tip}

## Ludwig Experiment Report

> Here are the results from the new model we have trained and validated. Using your knwoledge from the previous analysis, answers the following questions:

Reviewer: ```suggestion
Here are the results from the new model we have trained and validated. Using your knowledge from the previous analysis, answer the following questions:
```

>
>![alt](../../images/galaxy-ludwig/test2_entropy.png "Entropy - Epoch = 15")
>> <question-title></question-title>
>>
>> 1. For which label(s) did the model decrease the entropy? 
>>
>> 2. Why did the order of the labels change compared to the previous results?
>>
>>> <solution-title></solution-title>
>>>
>>> 1. The model decreased entropy across all labels. 
>>>
>>> 2. For some labels, the entropy drop was more pronounced, moving those labels toward the bottom of the list.
>>>
>>{: .solution}
>>
>{: .question}
>
{:  .comment}

> <comment-title>Loss</comment-title>
>
>![alt](../../images/galaxy-ludwig/test2_combined_loss.png "Combined Loss - Epoch = 15")
>> <question-title></question-title>
>>
>> 1. Did the increase in Epochs have a significant impact on the model's loss?
>>
>>> <solution-title></solution-title>
>>>
>>> 1. The model's training loss curve showed improvement, reaching a loss below 0.25 by the end of the epochs. 
>>>
>>{: .solution}
>>
>{: .question}
>
{:  .comment}

> <comment-title>Accuracy</comment-title>
>
>![alt](../../images/galaxy-ludwig/test2_accuracy.png "Accuracy - Epoch = 15")
>> <question-title></question-title>
>>

> >> 1. Did the increase in Epochs hava a significant impact on the model's accuracy 

Reviewer: ```suggestion
>> 1. Did the increase in Epochs have a significant impact on the model's accuracy 
```

>![recall formula](../../images/galaxy-ludwig/explain_recall_formula.png)
>
{: .tip}

> <tip-title>F1 score definition</tip-title>
> The F1 score is the harmonic mean of precision and recall. Unlike the simple average, the harmonic mean gives more weight to lower values, so the F1 score will be high only if both precision and recall are high.
>![F1 score formula](../../images/galaxy-ludwig/explain_f1_score.png)
>
{: .tip}

The F1 score vs. frequency label graph is a valuable tool for diagnosing how well a model handles different classes, especially in datasets where some classes are much more common than others. By analyzing this graph, you can identify areas where your model might need more attention, ensuring balanced and accurate predictions across all classes.

![F1 score vs frequency label](../../images/galaxy-ludwig/test2_f1_score.png "F1 Score vs. Frequency of Labels. This graph illustrates the relationship between the frequency of each class label in the dataset and the model's F1 score for that class. Higher F1 scores indicate better model performance.")

> Digits like '1' (occurrences > 7800 and F1 score > 0.992) and '0' (occurrences > 6800 and F1 score > 0.980) suggest that the model has learned to accurately identify these digits, despite of the difference when looking the frequency of the digits. Digit '5', with a lower frequency, still achieves a strong F1 score. Although its frequency is the lowest among the digits, the F1 score remains high, indicating that the model maintains a good balance of precision and recall even with fewer examples. This reflects the robustness of the CNN model.

Reviewer: ```suggestion
Digits like '1' (occurrences > 7800 and F1 score > 0.992) and '0' (occurrences > 6800 and F1 score > 0.980) suggest that the model has learned to accurately identify these digits, despite the difference when looking at the frequency of the digits. Digit '5', with a lower frequency, still achieves a strong F1 score. Although its frequency is the lowest among the digits, the F1 score remains high, indicating that the model maintains a good balance of precision and recall even with fewer examples. This reflects the robustness of the CNN model.
```

# Conclusion
In this tutorial, we demonstrated how to utilize the Galaxy-Ludwig tool to build an image recognition model using the MNIST dataset of handwritten digits. We followed a structured approach consisting of 
- uploading necessary files, 
- running the Ludwig experiment, and 
- evaluating the model's performance. 

> Through the process, we have seen how Ludwig simplifies the complexities of machine learning workflows, making it accessible and efficient for users. Additionally, we explored ways to enhance model performance by adjusting configurations and analyzing detailed evaluation metrics. 

Reviewer: ```suggestion
Through the process, we have seen how Ludwig simplifies the complexities of machine learning workflows, making them accessible and efficient for users. Additionally, we explored ways to enhance model performance by adjusting configurations and analyzing detailed evaluation metrics. 
```
